src/Microsoft.AspNet.Mvc.Core/HttpResponseStreamWriter.cs

```diff
@@ -90,59 +92,71 @@ public HttpResponseStreamWriter(Stream stream, Encoding encoding, int bufferSize
                 throw new ArgumentNullException(nameof(encoding));
             }
 
-            if (leasedByteBuffer == null)
-            {
-                throw new ArgumentNullException(nameof(leasedByteBuffer));
-            }
-
-            if (leasedCharBuffer == null)
+            if (bytePool == null)
             {
-                throw new ArgumentNullException(nameof(leasedCharBuffer));
+                throw new ArgumentNullException(nameof(bytePool));
             }
 
-            if (bufferSize <= 0 || bufferSize > leasedCharBuffer.Data.Count)
+            if (charPool == null)
             {
-                throw new ArgumentOutOfRangeException(nameof(bufferSize));
-            }
-
-            var requiredLength = encoding.GetMaxByteCount(bufferSize);
-            if (requiredLength > leasedByteBuffer.Data.Count)
-            {
-                var message = Resources.FormatHttpResponseStreamWriter_InvalidBufferSize(
-                    requiredLength,
-                    bufferSize,
-                    encoding.EncodingName,
-                    typeof(Encoding).FullName,
-                    nameof(Encoding.GetMaxByteCount));
-                throw new ArgumentException(message, nameof(leasedByteBuffer));
+                throw new ArgumentNullException(nameof(charPool));
             }
 
             _stream = stream;
             Encoding = encoding;
             _charBufferSize = bufferSize;
-            _leasedByteBuffer = leasedByteBuffer;
-            _leasedCharBuffer = leasedCharBuffer;
 
             _encoder = encoding.GetEncoder();
-            _byteBuffer = leasedByteBuffer.Data;
-            _charBuffer = leasedCharBuffer.Data;
+            _bytePool = bytePool;
+            _charPool = charPool;
+
+            try
+            {
+                _charBuffer = charPool.Rent(_charBufferSize);
```

Comment:
Could this be moved out of the try block? Is there a scenario where the assignment would fail. Ran into this when I was working with the view buffer changes.